File: blockrunnooor/router/decision.py
from __future__ import annotations

from dataclasses import dataclass

from ..models import Decision
from ..state.db import CircuitState, WalletState


@dataclass(frozen=True)
class DecisionResult:
    decision: Decision
    reason: str | None = None


def decide(wallet: WalletState, circuit: CircuitState, now: int) -> DecisionResult:
    if wallet.status != "active":
        return DecisionResult(decision="deny", reason=f"wallet_status:{wallet.status}")
    if wallet.cooldown_until and wallet.cooldown_until > now:
        return DecisionResult(decision="deny", reason="wallet_cooldown")
    if wallet.spent_today_usd >= wallet.daily_budget_usd and wallet.daily_budget_usd > 0:
        return DecisionResult(decision="deny", reason="daily_budget_exhausted")
    if circuit.open_until and circuit.open_until > now:
        return DecisionResult(decision="deny", reason="channel_circuit_open")
    return DecisionResult(decision="blockrun", reason=None)

File: src/router/decision.ts
import type { CircuitRow, WalletRow } from "../state/repo"

export type DecisionResult =
  | { decision: "deny"; reason: string }
  | { decision: "blockrun" }

export function decide(wallet: WalletRow, circuit: CircuitRow, now: number): DecisionResult {
  if (wallet.status !== "active") return { decision: "deny", reason: `wallet_status:${wallet.status}` }
  if (wallet.cooldown_until && wallet.cooldown_until > now) return { decision: "deny", reason: "wallet_cooldown" }
  if (wallet.daily_budget_usd > 0 && wallet.spent_today_usd >= wallet.daily_budget_usd) return { decision: "deny", reason: "daily_budget_exhausted" }
  if (circuit.open_until && circuit.open_until > now) return { decision: "deny", reason: "channel_circuit_open" }
  return { decision: "blockrun" }
}

